### PR TITLE
Endpoint/CEP Ownership Fix

### LINKED ISSUE #19931: Endpoint contains wrong Pod IPv4 address
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We observed dropped traffic between Pods for which the corresponding NetworkPolicy actually should allow traffic. Upon investigation we figured out, that only one specific Pod (from a StatefulSet) as affected, and that the problem is caused by the fact that Cilium stored a wrong IPv4 address in the Pod's `CiliumEndpoint` object.

The CiliumEndpoint's `status.networking.addressing.ipv4` field contained a wrong IPv4 address (`10.8.142.14`), not matching the effective IPv4 address of the Pod (`10.8.2.32`).

We have not yet figured out whether we can reliably reproduce this, right now it looks more like this is caused by a (rare) race condition (maybe related to StatefulSet specifics). 

**Endpoint**

```yaml
apiVersion: cilium.io/v2
kind: CiliumEndpoint
metadata:
  creationTimestamp: '2022-05-23T11:39:49Z'
  generation: 2
  labels:
    app.kubernetes.io/instance: grafana-agent-metrics
    app.kubernetes.io/managed-by: grafana-agent-operator
    app.kubernetes.io/name: grafana-agent
    app.kubernetes.io/version: v0.23.0
    controller-revision-hash: grafana-agent-metrics-shard-1-788dbb9c87
    grafana-agent: grafana-agent-metrics
    nx-k8s-topology-id: 8a8f0140-b805-44db-9dfe-625c6b5df899
    operator.agent.grafana.com/name: grafana-agent-metrics
    operator.agent.grafana.com/shard: '1'
    operator.agent.grafana.com/type: metrics
    statefulset.kubernetes.io/pod-name: grafana-agent-metrics-shard-1-0
  name: grafana-agent-metrics-shard-1-0
  namespace: grafana-agent-system
  ownerReferences:
    - apiVersion: v1
      blockOwnerDeletion: true
      kind: Pod
      name: grafana-agent-metrics-shard-1-0
      uid: 296f1523-ceac-4906-b4e1-0b8f4c4d5c01
  resourceVersion: '456543044'
  uid: 1c88af6b-67c0-4043-85f7-b821165f00ab
  selfLink: >-
    /apis/cilium.io/v2/namespaces/grafana-agent-system/ciliumendpoints/grafana-agent-metrics-shard-1-0
status:
  encryption: {}
  external-identifiers:
    container-id: 1fd8732c92dbf8283a2ecd25908604c4db0420965ee7e4f0eaba1870002e017a
    k8s-namespace: grafana-agent-system
    k8s-pod-name: grafana-agent-metrics-shard-1-0
    pod-name: grafana-agent-system/grafana-agent-metrics-shard-1-0
  id: 873
  identity:
    id: 54072
    labels:
      - k8s:app.kubernetes.io/instance=grafana-agent-metrics
      - k8s:app.kubernetes.io/managed-by=grafana-agent-operator
      - k8s:app.kubernetes.io/name=grafana-agent
      - k8s:app.kubernetes.io/version=v0.23.0
      - >-
        k8s:io.cilium.k8s.namespace.labels.grafana-agent-system.tree.hnc.x-k8s.io/depth=0
      - >-
        k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=grafana-agent-system
      - >-
        k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/name=grafana-agent
      - >-
        k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/namespace=kube-system
      - k8s:io.cilium.k8s.namespace.labels.scheduling.nexxiot.com/fargate=false
      - k8s:io.cilium.k8s.policy.cluster=default
      - k8s:io.cilium.k8s.policy.serviceaccount=grafana-agent
      - k8s:io.kubernetes.pod.namespace=grafana-agent-system
  named-ports:
    - name: http-metrics
      port: 8080
      protocol: TCP
  networking:
    addressing:
      - ipv4: 10.8.142.14
    node: 10.8.3.174
  state: ready
```

**Pod**

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: grafana-agent-metrics-shard-1-0
  generateName: grafana-agent-metrics-shard-1-
  namespace: grafana-agent-system
  uid: 296f1523-ceac-4906-b4e1-0b8f4c4d5c01
  resourceVersion: '456543083'
  creationTimestamp: '2022-05-23T11:39:48Z'
  labels:
    app.kubernetes.io/instance: grafana-agent-metrics
    app.kubernetes.io/managed-by: grafana-agent-operator
    app.kubernetes.io/name: grafana-agent
    app.kubernetes.io/version: v0.23.0
    controller-revision-hash: grafana-agent-metrics-shard-1-788dbb9c87
    grafana-agent: grafana-agent-metrics
    nx-k8s-topology-id: 8a8f0140-b805-44db-9dfe-625c6b5df899
    operator.agent.grafana.com/name: grafana-agent-metrics
    operator.agent.grafana.com/shard: '1'
    operator.agent.grafana.com/type: metrics
    statefulset.kubernetes.io/pod-name: grafana-agent-metrics-shard-1-0
  annotations:
    kubectl.kubernetes.io/default-container: grafana-agent
    kubernetes.io/psp: k8s.privileged-host
  ownerReferences:
    - apiVersion: apps/v1
      kind: StatefulSet
      name: grafana-agent-metrics-shard-1
      uid: 116802d1-f8a2-43d1-863e-6de8e5ca8150
      controller: true
      blockOwnerDeletion: true
  selfLink: /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-1-0
status:
  phase: Running
  conditions:
    - type: Initialized
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:49Z'
    - type: Ready
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:54Z'
    - type: ContainersReady
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:54Z'
    - type: PodScheduled
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:48Z'
  hostIP: 10.8.4.230
  podIP: 10.8.2.32
  podIPs:
    - ip: 10.8.2.32
  startTime: '2022-05-23T11:39:49Z'
  containerStatuses:
    - name: config-reloader
      state:
        running:
          startedAt: '2022-05-23T11:39:52Z'
      lastState: {}
      ready: true
      restartCount: 0
      image: quay.io/prometheus-operator/prometheus-config-reloader:v0.47.0
      imageID: >-
        quay.io/prometheus-operator/prometheus-config-reloader@sha256:0029252e7cf8cf38fc58795928d4e1c746b9e609432a2ee5417a9cab4633b864
      containerID: >-
        containerd://ade201bd0787e0e4ae9aeff887a091fad689eeec70406c65918e63908eb6a328
      started: true
    - name: grafana-agent
      state:
        running:
          startedAt: '2022-05-23T11:39:52Z'
      lastState: {}
      ready: true
      restartCount: 0
      image: docker.io/grafana/agent:v0.23.0
      imageID: >-
        docker.io/grafana/agent@sha256:a0beeaa6642c69efa472d509be2e2cf97dcb1c7e74047cca59a9452bf068f763
      containerID: >-
        containerd://8535bdd6e890f8028dc90eb68dab2ab317ee8e615b90322b81d6ce8e125e1438
      started: true
  qosClass: Burstable
spec:
  ...
```

### Cilium Version

1.11.3

### Kernel Version

5.4.188-104.359.amzn2.x86_64

### Kubernetes Version

1.21 (v1.21.5-eks-9017834)

### Sysdump

**Note**: this is a ZIP compressed BZIP2 Tar archive (had to work the system, BZIP2 to reduce size and ZIP to make Github accept the file format).

[cilium-sysdump-20220524-105632.tar.bz2.zip](https://github.com/cilium/cilium/files/8761164/cilium-sysdump-20220524-105632.tar.bz2.zip)

### Relevant log output

```shell
> hubble  observe -f --verdict DROPPED
May 24 08:49:02.049: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:02.049: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:03.077: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:03.077: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:05.093: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:05.093: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
```

### Anything else?

**NetworkPolicy**

Just for completeness, the corresponding `NetworkPolicy` object:

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kube-state-metrics
  namespace: kube-system
  labels:
    kustomize.toolkit.fluxcd.io/name: kube-state-metrics
    kustomize.toolkit.fluxcd.io/namespace: kube-system
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/name: kube-state-metrics
  ingress:
    - ports:
        - protocol: TCP
          port: 8080
        - protocol: TCP
          port: 8081
      from:
        - podSelector:
            matchLabels:
              app.kubernetes.io/name: grafana-agent
          namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: grafana-agent-system
    - ports:
        - protocol: TCP
          port: 8080
        - protocol: TCP
          port: 8081
      from:
        - podSelector: {}
  egress:
    - ports:
        - protocol: TCP
          port: 443
        - protocol: TCP
          port: 6443
  policyTypes:
    - Ingress
    - Egress
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tommyp1ckles**
Fixes: #19931

```release-note
Fix bugs where ciliumendpoints for statefulset pods where being incorrectly overwritten/deleted
```

### CODE REVIEW

**Comment by tommyp1ckles** [comment]
/test-1.25-net-next

**Review by jrajahalme — APPROVED** [review]

File: pkg/k8s/watchers/endpointsynchronizer.go [hunk 1]
```diff
@@ -246,6 +249,14 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 				// If it fails it means the test from the previous patch failed
 				// so we can safely replace this node in the CNP status.
 				replaceCEPStatus := []k8s.JSONPatch{
+					// If the stored UID matches the one in the ciliumendpoint then
+					// this first patch is a no-op, otherwise the entire patch will
+					// not be applied as uid is immutable.
+					{
+						OP:    "replace",
```
Comment (aanm) [inline pkg/k8s/watchers/endpointsynchronizer.go]: Why don't we use the `OP: "test"` that tests for the equalness of `/metadata/uid` and `e.GetCiliumEndpointUID()`?

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by thejosephstevens** [comment]
Quick question, I just came from the [originating issue](https://github.com/cilium/cilium/issues/19931) for this and wanted to check. The summary `Fix bugs where ciliumendpoints for statefulset pods where being incorrectly overwritten/deleted` seems to indicate this is targeted at Statefulsets, but this issue is also manifesting for Daemonsets. I scanned the code and it seems to be generic, but wanted to check if this PR is expected to also fix the same issue for Daemonsets.

**Comment by tommyp1ckles** [comment]
> Quick question, I just came from the [originating issue](https://github.com/cilium/cilium/issues/19931) for this and wanted to check. The summary `Fix bugs where ciliumendpoints for statefulset pods where being incorrectly overwritten/deleted` seems to indicate this is targeted at Statefulsets, but this issue is also manifesting for Daemonsets. I scanned the code and it seems to be generic, but wanted to check if this PR is expected to also fix the same issue for Daemonsets.

Hi @thejosephstevens, yeah good catch. 

Specifying StatefulSet was probably a mistake (I'll rewrite the commit message), more generally it's just when a pod is restarted of the same namespace/name as an existing one.

As you mentioned in your comment, this issue manifests when either Cilium patches/deletes the wrong CiliumEndpoint after a restart due to a race condition. This PR should fix by being more strict resource UIDs so it should fix any similar issue.

I'll make sure I can do a similar reproduction/fix-validation but with DaemonSets to double check however.

File: pkg/k8s/watchers/endpointsynchronizer.go (hunk 1 above)
Comment (tommyp1ckles) [reply pkg/k8s/watchers/endpointsynchronizer.go]: Yeah, that makes sense 👍

**Comment by tommyp1ckles** [comment]
@thejosephstevens See my comment in the original issue thread: https://github.com/cilium/cilium/issues/19931#issuecomment-1332975346

**Review by aanm — APPROVED** [review]

File: pkg/k8s/watchers/endpointsynchronizer.go [hunk 2]
```diff
@@ -246,6 +249,14 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 				// If it fails it means the test from the previous patch failed
 				// so we can safely replace this node in the CNP status.
 				replaceCEPStatus := []k8s.JSONPatch{
+					// If the stored UID matches the one in the ciliumendpoint then
+					// this first patch is a no-op, otherwise the entire patch will
+					// not be applied as uid is immutable.
+					{
+						OP:    "test",
+						Path:  "/metadata/uid",
+						Value: e.GetCiliumEndpointUID(),
```
Comment (aanm) [inline pkg/k8s/watchers/endpointsynchronizer.go:265]: @tommyp1ckles out of curiosity, what happens if we are starting a cilium agent for which we don't know the CEP UID? For example a Cilium agent restores endpoints that have an empty CEP UID set.

File: pkg/k8s/watchers/endpointsynchronizer.go [hunk 3]
```diff
@@ -263,6 +274,11 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 					k8stypes.JSONPatchType,
 					createStatusPatch,
 					meta_v1.PatchOptions{})
+				if err != nil {
+					scopedLog.WithError(err).Error("failed to update ciliumendpoint status")
+					return err
```
Comment (aanm) [inline pkg/k8s/watchers/endpointsynchronizer.go]: Just a FYI that I don't think we will receive an error in case the test for `/metadata/uid` fails

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/endpointsynchronizer.go (hunk 2 above)
Comment (tommyp1ckles) [reply pkg/k8s/watchers/endpointsynchronizer.go:265]: This situation is captured here: https://github.com/cilium/cilium/pull/21768/files#diff-6c96962323677414649840871089f13cad859f12b2b42ca62bcf317b49f69571R319

The idea is, that we can safely "take ownership" of a CEP when restoring from an empty UID reference only if that CEP was created locally.

File: pkg/k8s/watchers/endpointsynchronizer.go (hunk 3 above)
Comment (tommyp1ckles) [reply pkg/k8s/watchers/endpointsynchronizer.go]: It appears that if the UID doesn't match on the test this does return `k8serrors.StatusReasonInvalid`.

**Comment by tommyp1ckles** [comment]
/test-1.16-4.9

**Comment by tommyp1ckles** [comment]
/test-1.26-net-next

File: pkg/k8s/watchers/endpointsynchronizer_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+				ObjectMeta: v1.ObjectMeta{UID: types.UID("a")},
+				Status: v2.EndpointStatus{
+					Networking: &v2.EndpointNetworking{
+						NodeIP: "<nil>", // in testing, the node ip is returned as "<nil>"
+					},
+				},
+			},
+		},
+	} {
+		//ep := &endpoint.Endpoint{}
```
Comment (gandro) [inline pkg/k8s/watchers/endpointsynchronizer_test.go]: leftover comment?

**Review by aanm — APPROVED** [review]

File: pkg/k8s/watchers/endpointsynchronizer_test.go (hunk 4 above)
Comment (tommyp1ckles) [reply pkg/k8s/watchers/endpointsynchronizer_test.go]: yup, fixed

**Review by gandro — APPROVED** [review]

**Comment by tommyp1ckles** [comment]
@nathanjsweet Do you mind taking a look at this one as well 🙏

**Comment by tommyp1ckles** [comment]
@aanm Can we backport these changes?

**Review by nathanjsweet — APPROVED** [review]

**Comment by aanm** [comment]
> @aanm Can we backport these changes?

@tommyp1ckles let's do it for 1.13 and 1.12 for now /cc @joestringer

**Pull request merged**